apps/web/src/components/Messages/PreviewList.tsx

```diff
@@ -34,10 +35,36 @@ const PreviewList: FC<Props> = ({ className, selectedConversationKey }) => {
   const selectedTab = useMessageStore((state) => state.selectedTab);
   const setSelectedTab = useMessageStore((state) => state.setSelectedTab);
   const [showSearchModal, setShowSearchModal] = useState(false);
-  const { authenticating, loading, messages, profilesToShow, requestedCount, profilesError } =
-    useMessagePreviews();
+  const [currentIndex, setCurrentIndex] = useState<number>(0);
+  const { authenticating, conversationsLoading } = useListConversations();
+  const { messages, profilesError, previewLoading, profilesToShow, requestedCount, hasMore } =
+    useMessagePreviews(currentIndex);
+
+  const observer = useRef<IntersectionObserver | null>(null);
+
   const clearMessagesBadge = useMessagePersistStore((state) => state.clearMessagesBadge);
 
+  const lastElementRef = useCallback(
+    // @ts-expect-error custom infinite scroll
+    (node) => {
+      if (previewLoading) {
+        return;
+      }
+      if (observer.current) {
+        observer.current.disconnect();
+      }
+      observer.current = new IntersectionObserver((entries) => {
+        if (entries[0].isIntersecting && hasMore) {
+          setCurrentIndex(currentIndex + 20);
```

Comment:
Might be worth pulling this out in a constant since it's used in 2 different files.